### PR TITLE
feat: implement the logic for running python stored programs

### PR DESCRIPTION
**Pull request by sabhas**
## Intent

* add the ability the execute python stored program

## Implementation

* added a new runtime for python

## Checks

- [ ] Code is formatted correctly (`npm run lint:fix`).
- [ ] Any new functionality has been unit tested.
- [ ] All unit tests are passing (`npm test`).
- [ ] All CI checks are green.
- [ ] Reviewer is assigned.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -64,12 +64,22 @@ Example contents of a `.env` file:
 # Server mode is multi-user and suitable for intranet / internet use
 MODE=
 
+# A comma separated string that defines the available runTimes.
+# Priority is given to the runtime that comes first in the string.
+# Possible options at the moment are sas and js
+
+# options: [sas|js|py|sas,js|js,sas|py,sas|sas,py|py,js|js,py|sas,js,py|sas,py,js|js,sas,py|js,py,sas|py,sas,js|py,js,sas] default:sas
```
Comment (allanbowe) [inline README.md:71]: This list is going to get very long when we add R :-)

Comment (sabhas) [reply README.md:71]: Yeah, it definitely gonna be very long. So, what should we do? Should we skip these possible permutations of runtimes in readme?

Comment (allanbowe) [reply README.md:71]: Proposal:

```
# This string sets the priority of the available analytic runtimes
# Valid runtimes are SAS (sas), JavaScript (JS) and Python (py)
# For each option provided, there should be a corresponding path, 
# eg SAS_PATH or NODE_PATH
# Priority is given to runtimes earlier in the string
# Example options:  [sas,js,py | js,py | sas | sas,js]
```

File: api/src/controllers/internal/processProgram.ts [hunk 2]
```diff
@@ -47,6 +52,42 @@ export const processProgram = async (
       // copy the code.js program to log and end write stream
       writeStream.end(program)
 
+      session.completed = true
+      console.log('session completed', session)
```
Comment (YuryShkoda) [inline api/src/controllers/internal/processProgram.ts:56]: is this `console.log` left for a reason?

File: api/src/controllers/internal/processProgram.ts [hunk 3]
```diff
@@ -47,6 +52,42 @@ export const processProgram = async (
...
+    } catch (err: any) {
+      session.completed = true
+      session.crashed = err.toString()
+      console.log('session crashed', session.id, session.crashed)
```
Comment (YuryShkoda) [inline api/src/controllers/internal/processProgram.ts:60]: is this `console.log` left for a reason?

File: api/src/routes/api/spec/stp.spec.ts [hunk 4]
```diff
@@ -207,57 +208,61 @@ describe('stp', () => {
...
+          setupMocks()
+        })
+
+        afterEach(async () => {
+          jest.resetAllMocks()
+          await deleteFolder(path.join(filesFolder, testFilesFolder))
+        })
+
+        it('should execute python program when both python and sas program are present', async () => {
+          const programPath = path.join(testFilesFolder, 'program')
```
Comment (YuryShkoda) [inline api/src/routes/api/spec/stp.spec.ts:245]: I can see lots of duplications for each runtime, so it makes sense to create utils

**Review by YuryShkoda — CHANGES_REQUESTED** [review]
see comments above

File: README.md (hunk 1 above)
Comment (sabhas) [reply README.md]: updated

File: api/src/routes/api/spec/stp.spec.ts (hunk 4 above)
Comment (sabhas) [reply api/src/routes/api/spec/stp.spec.ts]: updated

File: api/src/controllers/internal/processProgram.ts (hunk 2 above)
Comment (sabhas) [reply api/src/controllers/internal/processProgram.ts:56]: Yes, we are intentionally using it here to log the status of the session. In the future, we will use some other logger then we'll replace it with that.

File: api/src/controllers/internal/processProgram.ts (hunk 3 above)
Comment (sabhas) [reply api/src/controllers/internal/processProgram.ts:60]: Yes, we are intentionally using it here to log the status of the session. In the future, we will use some other logger then we'll replace it with that.

File: api/src/controllers/internal/createPythonProgram.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+# change working directory to  session folder
+os.chdir(_SASJS_SESSION_PATH)
+
+# actual job code
+${program}
+
+`
+  // if no files are uploaded filesNamesMap will be undefined
+  if (otherArgs?.filesNamesMap) {
+    const uploadJSCode = await generateFileUploadPythonCode(
```
Comment (allanbowe) [inline api/src/controllers/internal/createPythonProgram.ts:56]: Perhaps `uploadJSCode` should be renamed?

File: api/src/controllers/internal/createPythonProgram.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+      otherArgs.filesNamesMap,
+      session.path
+    )
+
+    //If js code for the file is generated it will be appended to the top of jsCode
```
Comment (allanbowe) [inline api/src/controllers/internal/createPythonProgram.ts:61]: suggest:
```
  // If any files are uploaded, the program needs to be updated with some
  // dynamically generated variables (pointers) for ease of ingestion
```

Comment (sabhas) [reply api/src/controllers/internal/createPythonProgram.ts]: updated

File: api/src/controllers/internal/createPythonProgram.ts (hunk 5 above)
Comment (sabhas) [reply api/src/controllers/internal/createPythonProgram.ts]: renamed

**Review by allanbowe — APPROVED** [review]

**Pull request merged**